### PR TITLE
[SPARK-21485][FOLLOWUP][SQL][DOCS] Describes examples and arguments separately, and note/since in SQL built-in function documentation

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

This PR proposes to separate `extended` into `examples` and `arguments` internally so that both can be separately documented and add `since` and `note` for additional information.

For `since`, it looks users sometimes get confused by, up to my knowledge, missing version information. For example, see https://www.mail-archive.com/user@spark.apache.org/msg64798.html

For few good examples to check the built documentation, please see both:
`from_json` - https://spark-test.github.io/sparksqldoc/#from_json
`like` - https://spark-test.github.io/sparksqldoc/#like

For `DESCRIBE FUNCTION`, `note` and `since` are added as below:

```
> DESCRIBE FUNCTION EXTENDED rlike;
...
Extended Usage:
    Arguments:
      ...

    Examples:
      ...

    Note:
      Use LIKE to match with simple string pattern
```

```
> DESCRIBE FUNCTION EXTENDED to_json;
...
    Examples:
      ...

    Since: 2.2.0
```

For the complete documentation, see https://spark-test.github.io/sparksqldoc/

## How was this patch tested?

Manual tests and existing tests. Please see https://spark-test.github.io/sparksqldoc

Jenkins tests are needed to double check

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: I added `since` for both `from_json` and `to_json` as examples here. These were added from 2.2.0 - (https://issues.apache.org/jira/browse/SPARK-19637 and https://issues.apache.org/jira/browse/SPARK-19967)

Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala]: I added `since` for both `from_json` and `to_json` as examples here. These were added into SQL since 2.2.0 - (https://issues.apache.org/jira/browse/SPARK-19637 and https://issues.apache.org/jira/browse/SPARK-19967)

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java
Comment (HyukjinKwon) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java]: I believe we are okay to change this - https://github.com/apache/spark/blob/425c4ada4c24e338b45d0e9987071f05c5766fa5/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/package.scala#L22 and I could not find this in documentation.

**Comment by HyukjinKwon** [comment]
cc @rxin, @srowen and @cloud-fan, I believe this one is ready for a review. Could you take a look when you have some time?

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java
Comment (srowen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java]: There aren't many `ExpressionInfo` objects in memory right? adding more info to this bean doesn't have any meaningful performance implications, I presume. I suppose it's just breaking down the existing info further.

I also presume this is considered an internal API so it's OK to change the constructor. You could even retain the one constructor that is removed, just in case.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java
Comment (srowen) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java]: Agree, that's my only question, whether this change matters, because it's a developer API. You provide default implementations, though `extended()` gets removed. Hm. I am wondering if it's possible to keep `extended()` but, well, ignore it? it would at least be compatible even if it meant someone's implementation out there would have to update to provide information to `ExpressionInfo` correctly. That's not really a functional problem though.

**Comment by HyukjinKwon** [comment]
Will address comments within few days. (I am reading docs just to get used to things around my updated status)

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java
Comment (HyukjinKwon) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionInfo.java]: Yes, I think so for the former and sure will keep the original constructor.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java
Comment (HyukjinKwon) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java]: Will give a try.

File: sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java
Comment (HyukjinKwon) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java]: I manually tested source compatibility and I guess it also won't break binary compatibility - https://stackoverflow.com/questions/21197255/is-adding-a-method-to-a-java-annotation-safe-for-backward-compatibility

Comment (HyukjinKwon) [inline sql/catalyst/src/main/java/org/apache/spark/sql/catalyst/expressions/ExpressionDescription.java]: In the last commit https://github.com/apache/spark/pull/18749/commits/bf48875057ca5978b88cacab455292b64daf1557, I tried to keep the original behaviour and compatibility (although I guess we are fine to break this as said in few comments above):

For `ExpressionDescription` and `ExpressionInfo`,


if `extended` is an empty string (the default value), it creates the `extended` made by `arguments`, `examples`, `note` and `since` and use this as the extended description. 

if `extended` is a non-empty string, it ignores the new `arguments`, `examples`, `note` and `since` but use this as the extended description, assuming this was set manually.

**Comment by HyukjinKwon** [comment]
For source compatibility, in the commit, https://github.com/apache/spark/pull/18749/commits/389fc6ef788bf971f846ca36f49cea6a1c98b0d0, I tried to move back to `extended` to show it passes the tests and keep the source compatibility.

For binary compatibility, I tested in the way I described before - https://github.com/databricks/scala-style-guide/issues/46.

With `ExpressionDescription.java` assuming this is Spark 2.2.0:

```java
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
@Retention(RetentionPolicy.RUNTIME)
public @interface ExpressionDescription {
    String usage() default "";
    String extended() default "";
}
```

With `Main.scala` assuming this is the user application compiled with Spark 2.2.0:

```scala
@ExpressionDescription(
  usage = "I am usage",
  extended = """
    I am extended
  """)
case class A()

object Main {
  def main(args: Array[String]): Unit = {
    val df = scala.reflect.classTag[A].runtimeClass.getAnnotation(classOf[ExpressionDescription]).
    println(df.usage())
    println(df.extended())
  }
}
```

```
javac ExpressionDescription.java
scalac Main.scala
scala Main
```

prints

```
I am usage

    I am extended
```

And then, I changed `ExpressionDescription.java` assuming we merged this change into Spark's next release:

```diff
 import java.lang.annotation.Retention;
 import java.lang.annotation.RetentionPolicy;
 @Retention(RetentionPolicy.RUNTIME)
 public @interface ExpressionDescription {
     String usage() default "";
     String extended() default "";
+    String arguments() default "";
+    String examples() default "";
+    String note() default "";
+    String since() default "";
 }
```

and then I ran

```
javac ExpressionDescription.java
scala Main
```

this looks working fine as below:

```
I am usage

    I am extended

```

**Comment by HyukjinKwon** [comment]
I will revert 389fc6e if the test passes.

**Comment by HyukjinKwon** [comment]
The test failure above is due to `extended` being `null`. I also allowed this case in f76af78415085cff1f1d6dd31cb97d464b4fa52b.

**Comment by HyukjinKwon** [comment]
Now I am pretty confident for the compatibility concern given:

 - the commit, https://github.com/apache/spark/commit/389fc6ef788bf971f846ca36f49cea6a1c98b0d0 (and with a little additional change, 1fa02c8), where I intendedly left `extended`s as are in `ExpressionDescription` and `ExpressionInfo` usage, instead of setting new `examples` and `arguments` by the change proposal here to check the source compatibility.

- the manual tests, https://github.com/apache/spark/pull/18749#issuecomment-319286575, above to check binary compatibility. 

I am quite sure if the current state should be ready for another look if the last test passes, where I reverted the former, 389fc6e, above. I guess it should pass because the only diff is 1fa02c8 but will wait for the results before pinging again.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
@srowen, @rxin and @cloud-fan, I am confident for the compatibility concern per https://github.com/apache/spark/pull/18749#issuecomment-319364607 and believe it is ready for another look.

**Comment by rxin** [comment]
What is the compatibility concern?

**Comment by HyukjinKwon** [comment]
Here, both comments by Sean, https://github.com/apache/spark/pull/18749#discussion_r129929153 and https://github.com/apache/spark/pull/18749#discussion_r129927520.

It looks we are okay to change this given https://github.com/apache/spark/pull/18749#discussion_r129802019 but it is said in the comment, "DeveloperApi". So, I tried to keep the compatibility at my best.

So, I tried to think both possible compatibility cases:

- Checking If existing usage by developer still works with this change:

  ```java
   @ExpressionDescription(
      usage = "...",
      extended = """
        ...
      """)
  ```

  The current PR changes `extended` to `examples` and `arguments` in its usage. So, I manually reverted this usage and checked if it passes the test.

- Checking binary compatibility by annotation. I added few element in this PR, `examples`, `arguments`, `note` and `since`.  I manually checked this in https://github.com/apache/spark/pull/18749#issuecomment-319286575.

Did I maybe misunderstand something?

**Comment by rxin** [comment]
OK great. I think we should avoid breaking developer APIs, unless it has a huge upside. It wouldn't be fun to break it just for some cosmetic things ...

**Comment by HyukjinKwon** [comment]
Will keep in mind that.

**Review by srowen — APPROVED** [review]

**Comment by HyukjinKwon** [comment]
@rxin, does this look okay to you too?

**Comment by rxin** [comment]
seems fine to me.

**Comment by HyukjinKwon** [comment]
Will go merging this one. Thanks for reviewing this @srowen and @rxin.

**Comment by gatorsmile** [comment]
Before merging the PR, you have to wait for another committer to say LGTM. I think this is still required.

**Comment by rxin** [comment]
I think @srowen did it here using the new github approval: srowen approved these changes 20 hours ago

@srowen might be better if we stick with the LGTM one.

**Comment by gatorsmile** [comment]
https://spark-prs.appspot.com does not show GITHUB approval. If we can put LGTM explicitly, it will help us track the progress.

**Comment by HyukjinKwon** [comment]
I think we should rather make https://spark-prs.appspot.com recognising the github approval as well .. I considered github approval as an approval for this patch.

BTW, for now, is it documented that "LGTM" is required somewhere?

**Comment by rxin** [comment]
It is documented: http://spark.apache.org/contributing.html

It's been the convention forever and it's also good to use one way rather than multiple, so I'd prefer us just using that ... until we have a compelling reason to change.

**Comment by rxin** [comment]
Yes.

**Comment by HyukjinKwon** [comment]
I understood Spark have used LGTM as convention and it is a good way to show an approval as a sign-off but I meant, is "LGTM", not any other methods, required before merging it?

**Comment by rxin** [comment]
What's your point? You should be able to merge PR without anybody reviewing?

**Comment by HyukjinKwon** [comment]
I was wondering if other methods such as SGTM or a Github approval aa an approval for a patch are not allowed by rule. I usually say based on documentation or references to other guys and I read documentation a lot really but it wasn't in my head.. Just want to be sure and follow the rule. I am asking this really only because I am curious and want to be sure.

**Comment by rxin** [comment]
Ah OK. That's what we are discussing here. In the past it has always been an explicit "LGTM". That was defined before github had even the approval feature. Now most committers are actually not ASF github org members so they don't even have the permission to use this approve feature, and for whatever legacy reason almost everybody I know (except Sean) uses explicit LGTM. We also have tooling that checks the LGTM.

So rather than complicating tooling or changing everybody's behavior, I was suggesting to @srowen maybe he should just use LGTM.

**Comment by HyukjinKwon** [comment]
I see. Thanks for the details and explanation.

**Comment by srowen** [comment]
Formally, our model is "Commit Then Review". No approval or vote is required for any change, but changes can be retroactively vetoed. https://www.apache.org/foundation/glossary.html#CommitThenReview

Of course, we don't actually just commit things and then see what people think. But voting on every change is infeasible. It is a judgment call. My decision depends on things like: do I believe in good faith that any committer would be surprised if I merged this? are they likely to want to review first? am I confident enough to take on the consequence of fixing things if this causes a problem? Hence I leave changes open longer and ping more people the bigger the change is. Eventually, you get feedback from everyone who cares and tacit consent from anyone who hasn't commented.

Can you merge a change that nobody else has explicitly approved? yes, but, would only do so if it's trivial or if you have asked for a while for input and nobody has given any.

"LGTM" is not a formal mechanism or vote, and that text doesn't say it is. It's there to help contributors understand that's a common way to indicate approval.

I didn't realize that only ASF members can use the "Approve/Comment/Request Changes" feature. I prefer it just because it renders more nicely and doesn't generate an email. I see no reason to prefer that or "LGTM" or "SGTM" or "+1" or whatever. Any clear indication of approval is equivalent; these are just English words, not a formal vote. I am happy to update the doc to that effect.

**Comment by rxin** [comment]
Actually Sean I disagree. Spark has always been review then commit from the days before it entered ASF. In a huge debate last year within the ASF on RTC vs CTR, Spark was cited as a prominent example. Sometimes there are tiny teeny patches that committers can use their own judgements and don't necessarily require somebody else to sign-off.

**Comment by srowen** [comment]
RTC means a vote happens for each change: https://www.apache.org/foundation/glossary.html#ReviewThenCommit
That's not what we do. What debate are you referring to?

**Comment by HyukjinKwon** [comment]
(BTW, just for clarification, I think anyone can use that approve feature. I did it before - `https://github.com/apache/spark/pull/17734`)

**Comment by rxin** [comment]
@srowen search for "RTC vs CTR (was: Concerning Sentry...)"

From Todd Lipcon:
```
I don't have incubator stats... nor do I have a good way to measure "most
active" or "most successful" projects in the ASF (seems that itself could
be a 'centithread'-worthy discussion). But a potential proxy could be the
number of stars on github:
https://github.com/search?utf8=%E2%9C%93&q=user%3Aapache&type=Repositories&ref=searchresults
 (sort by number of stars)

Of the top ten:

Spark: RTC via github pull request
Storm: RTC (https://storm.apache.org/documentation/BYLAWS.html see "Code
Change")
Cassandra: RTC (based on my skimming the commit log which has "Reviewed by"
quite often)
CouchDB: RTC (http://couchdb.apache.org/bylaws.html see "RTC" section)
Kafka: RTC (based on "Reviewed by" showing up in recent commit logs)
Thrift: CTR
Mesos: RTC (based on reviewboard links in most of the recent commits)
Zookeeper: RTC (based on personal experience and comments above in this
thread)
Cordova: CTR (based on
https://github.com/apache/cordova-coho/blob/master/docs/committer-workflow.md
)
Hadoop: RTC (based on personal experience)

Briefly looking through the #11 through #30 projects I also see a
substantial number which operate on RTC (and others for which I don't know)

So, I don't think there's much evidence that RTC prevents a project from
becoming successful in the eyes of the developer community. Also worth
noting that several of these are relatively new TLPs (i.e. within the last
~3 years) whereas others are quite old but still active and successful.
```

BTW I didn't realize (and neither did Todd) ASF's definition of "RTC" is min 3 vote consensus, which is pretty odd.

The convention has always been find somebody who understands the area to review and give explicit LGTM. There are some exceptions like tiny changes (basically very similar to your paragraph on judgement call).

**Comment by rxin** [comment]
@HyukjinKwon you weren't a committer before :)

**Comment by srowen** [comment]
Yes, this isn't RTC. It isn't even RTC with 1-vote consensus, because you rightly say that you could merge with no other votes if it were obviously not required. But whatever. We can pick the process we want, and the name isn't important. I agree with your take and I think that's already the current process.

This change was clearly reviewed by several other people, so I don't see an issue with merging. If the objection was, "you didn't literally write 'LGTM'" then no I don't agree that this is important, and not a convention I've seen observed consistently. Think about the implication: someone says "yeah this looks good" and then you block for a day begging them to come back and write 'LGTM'?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by gatorsmile** [comment]
We have to define a clear rule. We do not want to revert the PRs merged by others. It just makes others awkward. 

In the past, based on my observation in the Spark SQL, when the committers left `SGTM` or `looks good to me`, it does not mean the quality is good enough for merging. That only means the quality or solution looks OK but needs more reviews/changes. I am not sure how the other Spark components work. However, I believe this is the rule we follow in Spark SQL. That is why I am surprised some committers merge the PR without an explicit LGTM message.

**Comment by srowen** [comment]
We have never had such a rule here or in any other project. Let's use common sense. If someone says 'looks good' that's fine. Or anything equivalent. If in doubt ask.

**Comment by gatorsmile** [comment]
In Spark SQL, we require such an explicit message. That is my understanding. If we want a change, I think we need to send a note to all the committers and let them know it. Then, they will leave such messages more carefully. 

Spark SQL becomes more and more complex. It is very easy to break the backward compatibility or even return an incorrect result. We are building infrastructure software. We have to be more careful, because our current test case coverage is still not good enough.

**Comment by HyukjinKwon** [comment]
@gatorsmile, I really don't want to go through your PRs and I have not ever done such things i the past. Let's document this and then let me follow. Please open a discussion in a mailing list if you are in doubt. We are not supposed to make a long term decision.

**Comment by srowen** [comment]
@gatorsmile you are the one asking for a change, and I do not see support for it. I am not aware of any Spark SQL-specific rule; what are you referring to? Reynold agrees above, for example, that in some cases it's justified to merge a change with no other review. This alone suggests there is no rule about typing "LGTM".

You rightly say that things need review. The important change, the more it needs review. That is not an argument for a voting rule; it's an argument for good judgment among committers, and I see no evidence that's lacking. An "LGTM" rule doesn't solve that problem: a committer with bad judgment can always approve something for merging that shouldn't, right? 

I further can't understand what's special about "LGTM". You are saying that committers say "sounds good to me" when they mean "this is not ready for merging" but whenever they say "LGTM" they always mean "this is ready for merging" -- sorry, what?

You are suggesting a change to a form of RTC, which would require a PMC vote. Consider the implications: nobody can merge a change until a vote completes. Are you sure? 

In the meantime, come now, let's move ahead with this PR, which was obviously reviewed.

**Comment by rxin** [comment]
@srowen that's not what I said. Almost always an explicit LGTM is preferred. There are tiny changes that might not require them, and it is up to the judgement of the committer. But those are more exceptions than the norm.

**Comment by srowen** [comment]
Yep, I agree with 'almost always', and tiny changes might not require them. I think that's what I said you said? "some cases". See also https://github.com/apache/spark/pull/18749#issuecomment-319892145

 But this alone disagrees with the idea that all changes need the string "LGTM" typed.

I am not sure why this popped up, but I propose we continue with the same process and conventions we always have. If anyone's unsure about whether an important review is affirmative: please ask.

Can we put this to bed now?

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by gatorsmile** [comment]
I am also afraid somebody might misunderstand what I or a few other committers say `SGTM` or `looks good` in Spark SQL. When a committer tries to merge Spark SQL related PRs based on these messages, please double checks. Does this sound reasonable?

**Comment by HyukjinKwon** [comment]
I think of opening a discussion in the mailing list if you are suggesting something. I don't think we should talk about this here.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
Other test failures look spurious and unrelated. I guess we are okay to go ahead with this PR anyway?

**Comment by gatorsmile** [comment]
LGTM.

Thanks! Merging to master.

**Pull request closed without merging**
